### PR TITLE
Opmphm Build Prediction Preparements

### PR DESCRIPTION
**Pull request by KurtMi**
# Purpose

* Introduces ks flag `KS_FLAG_REBUILD_OPMPHM`
* changes `opmphmIsBuild (...)` semantics
* adds the opmphm predictor
* adds a opmphm test case

# Checklist

Check relevant points but please do not remove entries.
For docu fixes, spell checking, and similar nothing
needs to be checked.

- [x] commit messages are fine ("module: short statement" syntax and refer to issues)
- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)
- [x] release notes are updated (doc/news/_preparation_next_release.md)

### CODE REVIEW

**Comment by KurtMi** [comment]
jenkins build all please

**Comment by KurtMi** [comment]
jenkins build all please

**Comment by KurtMi** [comment]
jenkins build all please

File: tests/ctest/test_opmphm_predictor.c [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	printf ("==================\n\n");
+
+	init (argc, argv);
+
+	test_internal_basic ();
+	test_internal_nochange ();
+	test_internal_change_whitebox ();
+	//~ test_ks ();
+
+	printf ("\ntest_opmphm_predictor RESULTS: %d test(s) done. %d error(s).\n", nbTest, nbError);
```
Comment (markus2330) [inline tests/ctest/test_opmphm_predictor.c]: We now have a print_result

File: tests/ctest/test_opmphm_predictor.c [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	test_internal_change_whitebox_set_to_state (3, 3, op, n);
+	test_internal_change_whitebox_set_to_state (3, 2, op, n);
+	test_internal_change_whitebox_set_to_state (2, 1, op, n);
+	test_internal_change_whitebox_set_to_state (1, 0, op, n);
+	test_internal_change_whitebox_set_to_state (0, 0, op, n);
+
+	opmphmPredictorDel (op);
+}
+
+//~ void test_ks (void)
```
Comment (markus2330) [inline tests/ctest/test_opmphm_predictor.c]: Why //~?

File: tests/ctest/test_opmphm_predictor.c [hunk 3]
```diff
@@ -0,0 +1,195 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see doc/LICENSE.md or https://www.libelektra.org)
+ */
+
+#include "../../src/libs/elektra/opmphmpredictor.c"
```
Comment (markus2330) [inline tests/ctest/test_opmphm_predictor.c]: Maybe its better to adopt the include path. `""` are also not really portable.

File: tests/ctest/test_opmphm.c [hunk 4]
```diff
@@ -20,6 +20,17 @@ const size_t maxComponentSize = 273;
...
 const uint8_t maxRUniPar = 10;
 
+static void test_basic (void)
+{
+	Opmphm * opmphm = opmphmNew ();
+	exit_if_fail (opmphm, "opmphmNew");
+	succeed_if (!opmphmIsBuild (opmphm), "isBuild");
+	succeed_if (!opmphmIsBuild (NULL), "isBuild");
+	opmphm->size = 1;
+	succeed_if (opmphmIsBuild (opmphm), "isBuild");
```
Comment (markus2330) [inline tests/ctest/test_opmphm.c]: very short message if it does not succeed.

File: src/libs/elektra/opmphmpredictor.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+void opmphmPredictorCopy (OpmphmPredictor * dest, OpmphmPredictor * source)
+{
+	ELEKTRA_NOT_NULL (source);
+	ELEKTRA_NOT_NULL (dest);
+	// copy values
+	dest->history = source->history;
+	dest->size = source->size;
+	dest->lookupCount = source->lookupCount;
+	// copy patternTable
+	memcpy (dest->patternTable, source->patternTable, source->size * sizeof (uint8_t));
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: there is a elektraMemcpy

File: src/libs/elektra/opmphmpredictor.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+ * @retval NULL memory error
+ */
+OpmphmPredictor * opmphmPredictorNew (void)
+{
+	OpmphmPredictor * out = elektraCalloc (sizeof (OpmphmPredictor));
+	if (!out)
+	{
+		return NULL;
+	}
+	uint16_t bytesInPatternTable = (OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK + 1) >> 2; // x >> 2 == (x * 2) / 8
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: Please give rationale for magic numbers.

File: src/libs/elektra/opmphmpredictor.c [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	// add the worth information to the history
+	op->history = ((op->history << 1) | wasItWorth) & OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK;
+	pos = op->history & OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK;
+	state = &(op->patternTable[pos >> 2]); // (pos * 2) / 8 == pos >> 2
+	// extract 2 bit state
+	uint8_t prediction = (*state >> ((pos & 0x3) << 1)) & 0x3; // ((pos & 0x3) << 1) == (pos % 4) * 2), 0x3 is 2 bit mask
+	// reset lookupCount
+	op->lookupCount = 1;
+	// determine prediction
+	if (prediction > 1)
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: :smile:

File: src/libs/elektra/opmphmpredictor.c [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+ * @retval 1 it is worth using the OPMPHM
+ * @retval 0 it is not worth using the OPMPHM
+ */
+inline int opmphmPredictorIncCountBinarySearch (OpmphmPredictor * op, size_t n)
+{
+	ELEKTRA_NOT_NULL (op);
+	++op->lookupCount;
+	size_t worthOpmphm = opmphmPredictorWorthOpmphm (n);
+	worthOpmphm = worthOpmphm + worthOpmphm + worthOpmphm; // 3 * worthOpmphm
+	if (op->lookupCount > worthOpmphm)
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: You can directly return a boolean :smile:

File: src/libs/elektra/opmphmpredictor.c [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+ *
+ * Tells how many ksLookup (...) invocations without alteration of the KeySet need to be made to justify the OPMPHM usage.
+ *
+ * @param n the number of elements in the KeySet
+ *
+ * @retval size_t the heuristic value
+ */
+inline size_t opmphmPredictorWorthOpmphm (size_t n)
+{
+	return n + 5000;
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: Please document magic numbers.

File: src/libs/elektra/opmphm.c [hunk 10]
```diff
@@ -589,8 +589,7 @@ int opmphmCopy (Opmphm * dest, const Opmphm * source)
  */
 int opmphmIsBuild (const Opmphm * opmphm)
 {
-	ELEKTRA_NOT_NULL (opmphm);
-	if (opmphm->size)
+	if (opmphm && opmphm->size)
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: please update docu of retval -1

File: src/libs/elektra/keyset.c [hunk 11]
```diff
@@ -1967,6 +1992,7 @@ static int elektraLookupBuildOpmphm (KeySet * ks)
 			return -1;
 		}
 	}
+	ELEKTRA_ASSERT (!opmphmIsBuild (ks->opmphm), "build already build OPMPHM");
```
Comment (markus2330) [inline src/libs/elektra/keyset.c]: sentence? "build already build OPMPHM"

File: src/libs/elektra/keyset.c [hunk 12]
```diff
@@ -1948,11 +1951,33 @@ static Key * elektraLookupBinarySearch (KeySet * ks, Key const * key, option_t o
...
+ * @internal
+ *
+ * @brief Builds the OPMPHM
+ *
+ * Creates the OPMPHM when not here.
+ * The passed KeySet must have a not build OPMPHM.
+ *
+ * @param ks the KeySet which OPMPHM is to build
+ *
+ * @return 0 on success
```
Comment (markus2330) [inline src/libs/elektra/keyset.c]: should be retval

File: src/libs/elektra/keyset.c [hunk 13]
```diff
@@ -1948,11 +1951,33 @@ static Key * elektraLookupBinarySearch (KeySet * ks, Key const * key, option_t o
...
+ * @return -1 on memory error or to many mapping invocations
```
Comment (markus2330) [inline src/libs/elektra/keyset.c]: also

File: src/include/kdbprivate.h [hunk 14]
```diff
@@ -133,11 +133,15 @@ typedef enum {
  * @ingroup backend
  */
 typedef enum {
-	KS_FLAG_SYNC = 1 /*!<
+	KS_FLAG_SYNC = 1, /*!<
 		 KeySet need sync.
 		 If keys were popped from the Keyset
 		 this flag will be set, so that the backend will sync
 		 the keys to database.*/
+	KS_FLAG_REBUILD_OPMPHM = 1 << 1 /*!<
```
Comment (markus2330) [inline src/include/kdbprivate.h]: should be within `#ifdef`.

File: src/include/kdbopmphmpredictor.h [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @brief Defines for the Order Preserving Minimal Perfect Hash Map Predictor.
+ *
+ * @copyright BSD License (see doc/COPYING or https://www.libelektra.org)
+ */
+#ifndef OPMPHM_PREDICTOR_H
+#define OPMPHM_PREDICTOR_H
+
+#include <stdint.h>
+#include <stdlib.h>
```
Comment (markus2330) [inline src/include/kdbopmphmpredictor.h]: Do we need this in the header file?

File: src/include/kdbopmphmpredictor.h [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * Maps the event of branch taken or not to a sequence of ksLookup (...) invocations
+ * without KeySet alteration that is worth using the OPMPHM or not.
+ */
+
+/**
+ * OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK defines the length and extraction mask of the global history register
+ * interpreted binary it must be a series of 1, with a minimum value of 0x3 and a maximum value of 0x7FFF.
+ */
+#define OPMPHM_PREDICTOR_HISTORY_EXTRACTION_MASK 0x7FF // 11 bit history
```
Comment (markus2330) [inline src/include/kdbopmphmpredictor.h]: const int?

File: doc/news/_preparation_next_release.md [hunk 17]
```diff
@@ -163,6 +163,7 @@ Thanks to Armin Wurzinger.
   of [`kdb ls`](https://master.libelektra.org/doc/help/kdb-ls.md). *(René Schwaiger)*
 - The documentation for `kdb` and `kdb set` now mention the `--` option to stop option processing. This is useful for setting negative values among other things. *(Klemens Böswirth)*
 - Plugins added with the flag `SHARED_ONLY` no longer get tested in the script `check_kdb_internal_check.sh` if executed with kdb-full or kdb-static. *(Armin Wurzinger)*
+- The OPMPHM has a new test case *(Kurt Micheli)*
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: This PR definitely does more

File: tests/ctest/test_opmphm_predictor.c (hunk 3 above)
Comment (KurtMi) [reply tests/ctest/test_opmphm_predictor.c]: How should I adopt the include path? Without `"` is is not working

File: src/libs/elektra/opmphmpredictor.c (hunk 5 above)
Comment (KurtMi) [reply src/libs/elektra/opmphmpredictor.c]: elektraMemcpy takes a Key as argument and the size is the per Key size:
`ssize_t elektraMemcpy (Key ** array1, Key ** array2, size_t size);`
I think it is not applicable for any `void *`.

File: tests/ctest/test_opmphm_predictor.c (hunk 3 above)
Comment (markus2330) [reply tests/ctest/test_opmphm_predictor.c]: https://cmake.org/cmake/help/v3.0/command/include_directories.html

`include_directories ("${CMAKE_SOURCE_DIR}/src/src/libs/elektra")` in the CMakeLists.txt and `#include <opmphmpredictor.c>` should work.

File: src/libs/elektra/opmphmpredictor.c (hunk 5 above)
Comment (markus2330) [reply src/libs/elektra/opmphmpredictor.c]: Yes, you are right.

**Comment by KurtMi** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Thank you! I am looking forward to the benchmarks!

**Pull request merged**